asmcomp/amd64/emit.mlp

```diff
@@ -916,213 +769,80 @@ let fundecl fundecl =
     if n <> 0
     then begin
       I.sub (int n) rsp;
-      cfi_adjust_cfa_offset n;
-    end;
-  end;
-  def_label !tailrec_entry_point;
-  emit_all true fundecl.fun_body;
-  List.iter emit_call_gc !call_gc_sites;
-  emit_call_bound_errors ();
-  if frame_required() then begin
-    let n = frame_size() - 8 - (if fp then 8 else 0) in
-    if n <> 0
-    then begin
-      cfi_adjust_cfa_offset (-n);
+      D.cfi_adjust_cfa_offset ~bytes:n;
     end;
-  end;
-  cfi_endproc ();
-  begin match system with
-  | S_gnu | S_linux ->
-      D.type_ (emit_symbol fundecl.fun_name) "@function";
-      D.size (emit_symbol fundecl.fun_name)
-        (ConstSub (
-            ConstThis,
-            ConstLabel (emit_symbol fundecl.fun_name)))
-  | _ -> ()
   end
 
-(* Emission of data *)
-
-let emit_item = function
-  | Cglobal_symbol s -> D.global (emit_symbol s)
-  | Cdefine_symbol s -> add_def_symbol s; _label (emit_symbol s)
-  | Cint8 n -> D.byte (const n)
-  | Cint16 n -> D.word (const n)
-  | Cint32 n -> D.long (const_nat n)
-  | Cint n -> D.qword (const_nat n)
-  | Csingle f -> D.long  (Const (Int64.of_int32 (Int32.bits_of_float f)))
-  | Cdouble f -> D.qword (Const (Int64.bits_of_float f))
-  | Csymbol_address s -> add_used_symbol s; D.qword (ConstLabel (emit_symbol s))
-  | Cstring s -> D.bytes s
-  | Cskip n -> if n > 0 then D.space n
-  | Calign n -> D.align n
-
-let data l =
-  D.data ();
-  List.iter emit_item l
+let fundecl decl =
+  Emitaux.fundecl decl
+    ~prepare
+    ~emit_all
+    ~emit_call
+    ~emit_jump_to_label
+    ~spacetime_before_uninstrumented_call
+    ~alignment_in_bytes:16
+    ~emit_numeric_constants:false
```

Comment:
having that as callbacks feels a bit strange. Wouldn't this looks better if these various functions where exposed in emit.mli, and Emitaux.fundecl is moved to another (new ?) unit higher in the dependency tree ?